New deployment workflow for releases

mislav opened the pull request:
- Kick off builds with `script/release <tagname>` instead of by pushing a git tag
- Build each platform (macOS, Linux, Windows) on their native OS
- macOS and Windows builds now have cgo enabled
- macOS builds are now signed and notarized
- The `checksums.txt` file attached to a release now includes the checksum of the MSI file too
- Ability to attach builds to an Actions workflow run instead of to a repository Release
- Build signing secrets are now protected behind a `production` deployment environment that needs to be manually approved each time
- Add ability to specify the Go version used for building (default: `1.19`)
- Add labels to assets attached to a GitHub Release

```sh
$ script/release --help

To tag a new release:
  script/release [--staging] <tag-name> [--platform {linux|macos|windows}] [--branch <branch>]

To build staging binaries from the current branch:
  script/release --current [--platform {linux|macos|windows}]

To build binaries locally with goreleaser:
  script/release --local --platform {linux|macos|windows}
```

It is no longer necessary to make pre-releases before cutting a stable release. If anything in the new release automation fails, the GitHub Release will have never been created and thus the git tag won't be created either. To test our release automation after changes, either use `script/release --staging` (note: this will skip signing) or `script/release vX.Y.Z-pre.N` (this will sign binaries and upload them to an actual prerelease GitHub Release).

⚠️ Removed:
- This stops updating our defunct scoop bucket https://github.com/cli/scoop-gh
- This stops moving project cards for https://github.com/cli/cli/projects/1

Followup items for after this ships:
- Extract building Debian/RPM repositories and publishing changes to cli.github.com to a separate job
- Figure out how to involve a step to provide a chance to amend the generated change log before publish

Closes https://github.com/cli/cli/issues/7404

samcoe commented on .github/workflows/deployment.yml:
```diff
@@ -0,0 +1,119 @@
+      - name: Configure GoReleaser
+        run: |
+          touch CHANGELOG.md
+          sed '/- id: windows$/,/^$/d; /- id: macos$/,/^$/d' .goreleaser.yml >.goreleaser.generated.yml
```
Rather than do this would it make more sense to have three goreleaser files checked in, one for each platform?

mislav replied:
I'm going to clean this approach up, but I still prefer having a single “master” Goreleaser config instead of syncing changes between multiple of them. Furthermore, a single Goreleaser config makes it easy to test cross-compilation for all platforms at once locally

mislav committed:
New deployment workflow

Add ability to trigger deployments without having to push a git tag

vilmibm commented on .github/workflows/deployment.yml:
```diff
@@ -1,25 +1,321 @@
         default: "1.19"
         type: string
+      platforms:
+        default: "linux,macos,windows"
```
can we at least get a comment here warning about the `sed` invocation?

vilmibm commented on .github/workflows/deployment.yml:
```diff
@@ -1,25 +1,321 @@
+      - name: Configure GoReleaser
+        run: |
+          touch CHANGELOG.md
+          sed '/#build:windows/,/^$/d; /#build:macos/,/^$/d' .goreleaser.yml >.goreleaser.generated.yml
```
this makes me uncomfortable but agreed that keeping three different files in sync isn't great either.

mislav replied:
I've moved goreleaser config generation into a `script/release` script where the hack is more comprehensible.

vilmibm reviewed (approved):
I don't really know how to give this a thorough review but it seems fine. we'll see how it shakes out for the next release.

Biswa96 committed:
Make: Fix target name for Windows platform (#7370)

This fixes the taget name by adding .exe extension for Windows platform.
Otherwise, the following error is shown with `make bin/gh.exe' command.

    make: *** No rule to make target 'script/build', needed by 'bin/gh.exe'.  Stop.

donatocayurin commented:
Thank you ❤️🙏

ananasik365 reviewed (commented):
L
